src/modules/Game.class.js

```diff
@@ -1,68 +1,265 @@
+        }
+
+        for (let i = 0; i < values.length; i++) {
+          if (values[i] === values[i + 1]) {
+            values[i] *= 2;
+            values[i + 1] = 0;
+            this.score += values[i];
+            isMovable = true;
+          }
+        }
+
+        const updatedRow = values.filter((value) => value !== 0);
+
+        while (updatedRow.length < 4) {
+          updatedRow.push(0);
+        }
+
+        for (let c = 0; c < 4; c++) {
+          if (this.state[r][c] !== updatedRow[3 - c]) {
+            this.state[r][c] = updatedRow[3 - c];
+            isMovable = true;
+          }
+        }
+      }
+
+      if (isMovable) {
+        this.getRandomCells();
+        this.checkGameStatus();
+      }
+    }
+  }
+  moveUp() {
+    if (this.status === Game.Status.playing) {
+      let isMovable = false;
+
+      for (let c = 0; c < 4; c++) {
+        const values = [];
+
+        for (let r = 0; r < 4; r++) {
+          if (this.state[r][c] !== 0) {
+            values.push(this.state[r][c]);
+          }
+        }
+
+        for (let i = 0; i < values.length; i++) {
+          if (values[i] === values[i + 1]) {
+            values[i] *= 2;
+            values[i + 1] = 0;
+            this.score += values[i];
+            isMovable = true;
+          }
+        }
+
+        const updatedColumn = values.filter((value) => value !== 0);
+
+        while (updatedColumn.length < 4) {
+          updatedColumn.push(0);
+        }
+
+        for (let r = 0; r < 4; r++) {
+          if (this.state[r][c] !== updatedColumn[r]) {
+            this.state[r][c] = updatedColumn[r];
+            isMovable = true;
+          }
+        }
+      }
+
+      if (isMovable) {
+        this.getRandomCells();
+        this.checkGameStatus();
+      }
+    }
   }
+  moveDown() {
+    if (this.status === Game.Status.playing) {
+      let isMovable = false;
 
-  moveLeft() {}
-  moveRight() {}
-  moveUp() {}
-  moveDown() {}
-
-  /**
-   * @returns {number}
-   */
-  getScore() {}
-
-  /**
-   * @returns {number[][]}
-   */
-  getState() {}
-
-  /**
-   * Returns the current game status.
-   *
-   * @returns {string} One of: 'idle', 'playing', 'win', 'lose'
-   *
-   * `idle` - the game has not started yet (the initial state);
-   * `playing` - the game is in progress;
-   * `win` - the game is won;
-   * `lose` - the game is lost
-   */
-  getStatus() {}
-
-  /**
-   * Starts the game.
-   */
-  start() {}
-
-  /**
-   * Resets the game.
-   */
-  restart() {}
-
-  // Add your own methods here
+      for (let c = 0; c < 4; c++) {
+        const values = [];
+
+        for (let r = 3; r >= 0; r--) {
+          if (this.state[r][c] !== 0) {
+            values.push(this.state[r][c]);
+          }
+        }
+
+        for (let i = 0; i < values.length; i++) {
+          if (values[i] === values[i + 1]) {
+            values[i] *= 2;
+            values[i + 1] = 0;
+            this.score += values[i];
+            isMovable = true;
+          }
```

Comment:
The same potential bug as in `moveLeft` method is present here. Ensure that after merging, the next cell is skipped to avoid multiple merges in one move.